fix: 🐛 implemented a circular dependency check in the ServiceRegistry

marleypowell opened the pull request:
fix: :bug: implemented a circular dependency check in the `ServiceRegistry`

fixes #19264

## Hey, I just made a Pull Request!

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Rugvip commented on packages/backend-app-api/src/wiring/ServiceRegistry.ts:
```diff
@@ -146,6 +146,43 @@ export class ServiceRegistry implements EnumerableServiceHolder {
+      depChain.push(node);
+    }
+
+    while (!!node && node.id !== head.service.id) {
```
You'll want to keep track of visited nodes, or you might end up with an infinite loop in the case where the starting node is not part of the cycle, for example `A -> B -> C -> B`.

marleypowell replied:
Added a unit test to cover this scenario

Rugvip commented on packages/backend-app-api/src/wiring/ServiceRegistry.ts:
```diff
@@ -146,6 +146,43 @@ export class ServiceRegistry implements EnumerableServiceHolder {
+      }
+      node = nodes.shift();
+      if (node) {
+        depChain.push(node);
```
Always pushing a node here seems like it would result in incorrect chains in many cases. For example `A -> B,C` and `C -> A`. I think that'd tell you that the chain is `A -> B -> C -> A` right?

marleypowell replied:
Added a unit test to cover this scenario

Rugvip reviewed (changes requested):
Thank you for taking a look at this! 👍 

I actually ended up implementing a similar check for modules over in https://github.com/backstage/backstage/pull/19337/files#diff-b5085d072500931193f6d660a80ad32abac0056a02a7865c157dd87030a81f36, what do you think about integrating that once it's merged instead?

marleypowell commented:
> Thank you for taking a look at this! 👍
> 
> I actually ended up implementing a similar check for modules over in https://github.com/backstage/backstage/pull/19337/files#diff-b5085d072500931193f6d660a80ad32abac0056a02a7865c157dd87030a81f36, what do you think about integrating that once it's merged instead?

As suggested I've updated `ServiceRegistry` to use `DependencyGraph` for circular dependency checks. Also as discussed I've fixed a small issue in `DependencyGraph` where the returned circular chain was in reverse order.

Rugvip commented on packages/backend-app-api/src/wiring/types.ts:
```diff
@@ -39,6 +39,7 @@ export interface CreateSpecializedBackendOptions {
 
 export interface ServiceHolder {
   get<T>(api: ServiceRef<T>, pluginId: string): Promise<T> | undefined;
+  checkForCircularDeps(): void;
```
Hmm, this doesn't quite make sense to add to any of these interfaces. Given that, I'm feeling it makes the most sense to keep `checkForCircularDeps` private and instead add a static `ServiceRegistry.create()` factory that instantiates the registry and checks for circular deps, removing the call from `BackendInitialized.start`

marleypowell replied:
Updated to have a static create method and private constructor. I've made the changeset a minor bump (IIRC that's fine for breaking changes pre v1.0.0)

Rugvip commented on .changeset/tricky-melons-fold.md:
```diff
@@ -0,0 +1,6 @@
+'@backstage/backend-app-api': patch
+---
+
+fix: :bug: implemented up front circular dependency check in the `BackendInitializer`
+fix: :bug: updated `detectCircularDependency` in `DependencyGraph` to return circular dependencies starting from the first node
```
GitHub emoji shorthands might not show up everywhere the changelogs are rendered, so these would need to be emoji characters instead. Tbh I think it's best to skip here though, changelogs are very serious business 😁

marleypowell replied:
Removed emoji references.

Rugvip reviewed (changes requested):
Sweet! 😁 

Just a few nits, and thank you for building out the test suite!

Rugvip reviewed (changes requested):
Thank you! 👍 

Just a changeset nit

marleypowell commented on .changeset/tricky-melons-fold.md:
```diff
@@ -0,0 +1,7 @@
+
+refactor!: updated `ServiceRegistry` to have a static create method and private constructor.
+fix: updated `ServiceRegistry` to perform circular dependency check on creation.
+fix: updated `detectCircularDependency` in `DependencyGraph` to return circular dependencies starting from the first node
```
thanks, applied suggestion.
